# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas Basics (DataFrame Basics I)

### Exercise 1: Importing and Inspecting your own Dataset

Before we can work with the Pandas Library, we need to __import Pandas__ (by convention as __pd__). <br>
1. __Import__ Pandas. __Fill in the gaps__!

In [ ]:
import pandas as pd

Now we want to import the __Cars Dataset__ that is stored in the CSV-File __cars.csv__. <br>
2. __Import__ the Cars Dataset and __store__ the Pandas DataFrame in the variable __cars__! __Fill in the gaps__!

In [ ]:
cars = pd.read_csv("datasets/cars.csv")

#### Some more details on the Columns:

- __mpg__: Measures __fuel efficiency__ in terms of miles per gallon. __Higher mpg__ means __higher efficiency__
- __cylinders__: Number of cylinders in the car
- __displacement__: Technical feature (engine)
- __horsepower__: Unit of measurement for engine power 
- __weight__: weight in lbs
- __acceleration__: time to accelerate from 0 to 60 mph (sec.)
- __model_year__: self explanatory
- __origin__: origin of the manufacturer/car
- __name__: model name

3a. __Inspect__ the DataFrame cars by "printing" cars!

In [ ]:
cars

3b. __Inspect__ the __first 10 Rows__ of the DataFrame cars! __Fill in the gaps__! What is the __name of the very first car__ in the first row (index label/pos = 0)?

In [ ]:
print(cars.iloc[0])
cars.head(10)

It´s a ... Chevrolet Chevelle Malibu!

3c. Also inspect the __last 5 Rows__! __How many cars/rows__ do we have in the Dataset (mind the range index starting from 0!)?

In [ ]:
print(f"Number of cars = {len(cars)}")
cars.tail(5)

We have ... 398 Cars!

4. __Inspect__ the entire DataFrame (__all rows__)! __Fill in the gaps!__

In [ ]:
pd.options.display.max_rows = 400

In [ ]:
cars

5. Get some __meta information__ on our DataFrame! In which __Column__ do we have __missing/NaN__ Values?

In [37]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


We have 6 missing/NaN Values in the column ... horsepower!

6. Let´s get some __summary statistics__ on our DataFrame! What is the __maximum__ value in the column __horsepower__?

In [47]:
print(cars["horsepower"].max())
cars.describe()

230.0


mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

The maximum value in the column horsepower is ... 230!

-------------------------------

# Well Done!

---------------------------------------------------------------------------------------------------------------------